File: services/risk-assessment-service/ml-training/data/synthetic_generator.py
"""
Synthetic Time-Series Risk Data Generator

Generates 2-3 years of historical risk data for LSTM training:
- Industry-specific default rate patterns
- Seasonal trends (quarterly, annual cycles)
- Economic cycle simulation (recession, growth periods)
- Random walk with drift for risk evolution
- 10,000+ business time-series (100-1000 per industry)
"""

import numpy as np
import pandas as pd
from typing import Dict, List, Tuple, Optional
import random
from datetime import datetime, timedelta
import yaml
import os


class SyntheticRiskDataGenerator:
    """Generates synthetic time-series risk data for LSTM training"""
    
    def __init__(self, config_path: str = "models/model_config.yaml"):
        """Initialize the generator with configuration"""
        self.config = self._load_config(config_path)
        self.industries = self.config.get('industries', [
            'technology', 'finance', 'healthcare', 'manufacturing', 
            'retail', 'energy', 'real_estate', 'transportation'
        ])
        self.sequence_length = self.config.get('sequence_length', 36)  # 3 years
        self.feature_count = self.config.get('feature_count', 20)
        self.businesses_per_industry = self.config.get('businesses_per_industry', 1250)  # 10k total
        
        # Set random seed for reproducibility
        np.random.seed(self.config.get('random_seed', 42))
        random.seed(self.config.get('random_seed', 42))
        
        # Industry-specific risk patterns
        self.industry_patterns = self._initialize_industry_patterns()
        
    def _load_config(self, config_path: str) -> Dict:
        """Load configuration from YAML file"""
        if os.path.exists(config_path):
            with open(config_path, 'r') as f:
                return yaml.safe_load(f)
        else:
            # Default configuration
            return {
                'sequence_length': 36,
                'feature_count': 20,
                'businesses_per_industry': 1250,
                'random_seed': 42,
                'industries': [
                    'technology', 'finance', 'healthcare', 'manufacturing', 
                    'retail', 'energy', 'real_estate', 'transportation'
                ]
            }
    
    def _initialize_industry_patterns(self) -> Dict:
        """Initialize industry-specific risk patterns"""
        patterns = {}
        
        for industry in self.industries:
            patterns[industry] = {
                'base_risk': np.random.uniform(0.1, 0.4),  # Base risk level
                'volatility': np.random.uniform(0.05, 0.2),  # Risk volatility
                'seasonality': {
                    'q1_factor': np.random.uniform(0.8, 1.2),
                    'q2_factor': np.random.uniform(0.9, 1.1),
                    'q3_factor': np.random.uniform(0.8, 1.2),
                    'q4_factor': np.random.uniform(0.7, 1.3)
                },
                'trend': np.random.uniform(-0.02, 0.02),  # Monthly trend
                'economic_sensitivity': np.random.uniform(0.5, 1.5)  # Economic cycle sensitivity
            }
        
        return patterns
    
    def generate_business_sequence(self, business_id: str, industry: str, 
                                 start_date: datetime) -> pd.DataFrame:
        """Generate a time-series sequence for a single business"""
        
        pattern = self.industry_patterns[industry]
        dates = [start_date + timedelta(days=30*i) for i in range(self.sequence_length)]
        
        # Initialize risk score with base industry risk
        risk_scores = np.zeros(self.sequence_length)
        risk_scores[0] = pattern['base_risk'] + np.random.normal(0, 0.05)
        
        # Generate time-series with trend, seasonality, and volatility
        for t in range(1, self.sequence_length):
            # Previous risk score
            prev_risk = risk_scores[t-1]
            
            # Trend component
            trend = pattern['trend']
            
            # Seasonal component
            month = dates[t].month
            quarter = (month - 1) // 3 + 1
            seasonal_factor = pattern['seasonality'][f'q{quarter}_factor']
            seasonal = (seasonal_factor - 1) * 0.1
            
            # Economic cycle component (simulated)
            economic_cycle = self._generate_economic_cycle(t)
            economic_impact = economic_cycle * pattern['economic_sensitivity'] * 0.1
            
            # Random walk component
            random_walk = np.random.normal(0, pattern['volatility'])
            
            # Calculate new risk score
            new_risk = prev_risk + trend + seasonal + economic_impact + random_walk
            
            # Ensure risk score stays within bounds [0, 1]
            risk_scores[t] = np.clip(new_risk, 0.01, 0.99)
        
        # Generate additional features
        features = self._generate_business_features(business_id, industry, risk_scores, dates)
        
        # Create DataFrame
        data = {
            'business_id': [business_id] * self.sequence_length,
            'industry': [industry] * self.sequence_length,
            'date': dates,
            'risk_score': risk_scores,
            'risk_level': [self._score_to_level(score) for score in risk_scores]
        }
        
        # Add generated features
        for i, feature_name in enumerate(features.columns):
            data[f'feature_{i}'] = features[feature_name].values
        
        return pd.DataFrame(data)
    
    def _generate_economic_cycle(self, time_step: int) -> float:
        """Generate economic cycle component (recession/growth periods)"""
        # Simulate economic cycles with 4-year periods
        cycle_period = 48  # 4 years in months
        cycle_phase = (time_step % cycle_period) / cycle_period * 2 * np.pi
        
        # Economic cycle: -1 (recession) to +1 (growth)
        economic_cycle = np.sin(cycle_phase)
        
        # Add some noise to make it more realistic
        economic_cycle += np.random.normal(0, 0.1)
        
        return economic_cycle
    
    def _generate_business_features(self, business_id: str, industry: str, 
                                  risk_scores: np.ndarray, dates: List[datetime]) -> pd.DataFrame:
        """Generate additional business features for the time-series"""
        
        features = pd.DataFrame(index=range(len(risk_scores)))
        
        # Financial metrics (correlated with risk)
        features['revenue_growth'] = np.random.normal(0.02, 0.1, len(risk_scores))
        features['profit_margin'] = np.random.uniform(0.05, 0.25, len(risk_scores))
        features['debt_ratio'] = np.random.uniform(0.2, 0.8, len(risk_scores))
        
        # Operational metrics
        features['employee_count'] = np.random.randint(10, 1000, len(risk_scores))
        features['customer_satisfaction'] = np.random.uniform(3.0, 5.0, len(risk_scores))
        features['market_share'] = np.random.uniform(0.01, 0.3, len(risk_scores))
        
        # Compliance and regulatory
        features['compliance_score'] = np.random.uniform(0.7, 1.0, len(risk_scores))
        features['regulatory_changes'] = np.random.poisson(0.5, len(risk_scores))
        
        # Market conditions
        features['market_volatility'] = np.random.uniform(0.1, 0.4, len(risk_scores))
        features['competition_intensity'] = np.random.uniform(0.3, 0.9, len(risk_scores))
        
        # Time-based features
        features['month'] = [d.month for d in dates]
        features['quarter'] = [(d.month - 1) // 3 + 1 for d in dates]
        features['year'] = [d.year for d in dates]
        
        # Lagged features (previous period values)
        for lag in [1, 3, 6, 12]:
            if lag < len(risk_scores):
                lagged_values = np.roll(risk_scores, lag)
                lagged_values[:lag] = risk_scores[0]
                features[f'risk_score_lag_{lag}'] = lagged_values
        
        # Moving averages
        for window in [3, 6, 12]:
            features[f'risk_score_ma_{window}'] = pd.Series(risk_scores).rolling(window=window, min_periods=1).mean()
        
        # Volatility measures
        features['risk_volatility_3m'] = pd.Series(risk_scores).rolling(window=3, min_periods=1).std()
        features['risk_volatility_6m'] = pd.Series(risk_scores).rolling(window=6, min_periods=1).std()
        
        # Ensure we have exactly the required number of features
        if len(features.columns) > self.feature_count:
            features = features.iloc[:, :self.feature_count]
        elif len(features.columns) < self.feature_count:
            # Add random features to reach target count
            for i in range(len(features.columns), self.feature_count):
                features[f'random_feature_{i}'] = np.random.normal(0, 1, len(risk_scores))
        
        return features
    
    def _score_to_level(self, score: float) -> str:
        """Convert risk score to risk level"""
        if score < 0.3:
            return 'low'
        elif score < 0.6:
            return 'medium'
        elif score < 0.8:
            return 'high'
        else:
            return 'critical'
    
    def generate_dataset(self, output_path: str = "data/synthetic_risk_data.parquet") -> pd.DataFrame:
        """Generate the complete synthetic dataset"""
        
        print(f"Generating synthetic dataset with {len(self.industries)} industries...")
        print(f"Target: {self.businesses_per_industry} businesses per industry")
        print(f"Sequence length: {self.sequence_length} months")
        print(f"Features per timestep: {self.feature_count}")
        
        all_data = []
        business_id_counter = 0
        
        for industry in self.industries:
            print(f"Generating data for {industry} industry...")
            
            for i in range(self.businesses_per_industry):
                business_id = f"business_{business_id_counter:06d}"
                
                # Random start date within the last 3 years
                start_date = datetime.now() - timedelta(days=365*3 + np.random.randint(0, 365))
                
                # Generate sequence for this business
                business_data = self.generate_business_sequence(business_id, industry, start_date)
                all_data.append(business_data)
                
                business_id_counter += 1
                
                if (i + 1) % 100 == 0:
                    print(f"  Generated {i + 1}/{self.businesses_per_industry} businesses")
        
        # Combine all data
        print("Combining all data...")
        dataset = pd.concat(all_data, ignore_index=True)
        
        # Add some data quality checks
        print("Performing data quality checks...")
        self._validate_dataset(dataset)
        
        # Save dataset
        print(f"Saving dataset to {output_path}...")
        os.makedirs(os.path.dirname(output_path), exist_ok=True)
        dataset.to_parquet(output_path, index=False)
        
        print(f"Dataset generation complete!")
        print(f"Total records: {len(dataset):,}")
        print(f"Unique businesses: {dataset['business_id'].nunique():,}")
        print(f"Date range: {dataset['date'].min()} to {dataset['date'].max()}")
        print(f"Industries: {dataset['industry'].nunique()}")
        
        return dataset
    
    def _validate_dataset(self, dataset: pd.DataFrame) -> None:
        """Validate the generated dataset"""
        
        # Check for missing values
        missing_values = dataset.isnull().sum().sum()
        if missing_values > 0:
            print(f"Warning: {missing_values} missing values found")
        
        # Check risk score distribution
        risk_stats = dataset['risk_score'].describe()
        print(f"Risk score statistics:")
        print(f"  Mean: {risk_stats['mean']:.3f}")
        print(f"  Std: {risk_stats['std']:.3f}")
        print(f"  Min: {risk_stats['min']:.3f}")
        print(f"  Max: {risk_stats['max']:.3f}")
        
        # Check risk level distribution
        risk_level_dist = dataset['risk_level'].value_counts(normalize=True)
        print(f"Risk level distribution:")
        for level, proportion in risk_level_dist.items():
            print(f"  {level}: {proportion:.1%}")
        
        # Check industry distribution
        industry_dist = dataset['industry'].value_counts()
        print(f"Industry distribution:")
        for industry, count in industry_dist.items():
            print(f"  {industry}: {count:,} records")


def main():
    """Main function to generate synthetic dataset"""
    
    # Create generator
    generator = SyntheticRiskDataGenerator()
    
    # Generate dataset
    dataset = generator.generate_dataset("data/synthetic_risk_data.parquet")
    
    # Print summary statistics
    print("\n" + "="*50)
    print("DATASET SUMMARY")
    print("="*50)
    print(f"Shape: {dataset.shape}")
    print(f"Memory usage: {dataset.memory_usage(deep=True).sum() / 1024**2:.1f} MB")
    print(f"Columns: {list(dataset.columns)}")


if __name__ == "__main__":
    main()

File: services/risk-assessment-service/internal/ml/data/synthetic_generator.go
package data

import (
	"fmt"
	"math"
	"math/rand"
	"time"

	"go.uber.org/zap"

	"kyb-platform/services/risk-assessment-service/internal/models"
)

// SyntheticDataGenerator generates synthetic time-series data for LSTM training and inference
type SyntheticDataGenerator struct {
	industryPatterns map[string]RiskPattern
	randomSeed       int64
	logger           *zap.Logger
}

// RiskPattern defines the risk characteristics for an industry
type RiskPattern struct {
	BaseRisk            float64
	Volatility          float64
	Seasonality         []float64 // 12 months of seasonal factors
	Trend               float64   // Annual trend
	EconomicSensitivity float64   // Sensitivity to economic cycles
}

// RiskDataPoint represents a single data point in the time series
type RiskDataPoint struct {
	Timestamp        time.Time
	RiskScore        float64
	FinancialHealth  float64
	ComplianceScore  float64
	MarketConditions float64
	RevenueTrend     float64
	EmployeeGrowth   float64
	RiskVolatility   float64
}

// NewSyntheticDataGenerator creates a new synthetic data generator
func NewSyntheticDataGenerator() *SyntheticDataGenerator {
	return &SyntheticDataGenerator{
		industryPatterns: initializeIndustryPatterns(),
		randomSeed:       time.Now().UnixNano(),
		logger:           zap.NewNop(),
	}
}

// SetLogger sets the logger for the synthetic data generator
func (sdg *SyntheticDataGenerator) SetLogger(logger *zap.Logger) {
	sdg.logger = logger
}

// GenerateHistoricalSequence generates a historical sequence of risk data for a business
func (sdg *SyntheticDataGenerator) GenerateHistoricalSequence(business *models.RiskAssessmentRequest, months int) []RiskDataPoint {
	// Get industry pattern
	pattern, exists := sdg.industryPatterns[business.Industry]
	if !exists {
		pattern = sdg.industryPatterns["default"]
	}

	// Set random seed for reproducible results
	rng := rand.New(rand.NewSource(sdg.randomSeed + int64(len(business.BusinessName))))

	// Generate sequence
	sequence := make([]RiskDataPoint, months)
	baseTime := time.Now().AddDate(0, -months, 0)

	for i := 0; i < months; i++ {
		currentTime := baseTime.AddDate(0, i, 0)
		monthIndex := int(currentTime.Month()) - 1

		// Generate base risk score with trend and seasonality
		trendFactor := float64(i) * pattern.Trend / 12.0
		seasonalFactor := pattern.Seasonality[monthIndex]
		baseRisk := pattern.BaseRisk + trendFactor + seasonalFactor

		// Add random walk component
		randomWalk := (rng.Float64() - 0.5) * pattern.Volatility
		riskScore := baseRisk + randomWalk

		// Ensure risk score is in valid range
		riskScore = math.Max(0.0, math.Min(1.0, riskScore))

		// Generate correlated features
		financialHealth := sdg.generateCorrelatedFeature(riskScore, 0.7, rng)
		complianceScore := sdg.generateCorrelatedFeature(riskScore, 0.6, rng)
		marketConditions := sdg.generateMarketConditions(currentTime, rng)
		revenueTrend := sdg.generateCorrelatedFeature(riskScore, 0.5, rng)
		employeeGrowth := sdg.generateCorrelatedFeature(riskScore, 0.4, rng)
		riskVolatility := sdg.generateVolatility(riskScore, pattern.Volatility, rng)

		sequence[i] = RiskDataPoint{
			Timestamp:        currentTime,
			RiskScore:        riskScore,
			FinancialHealth:  financialHealth,
			ComplianceScore:  complianceScore,
			MarketConditions: marketConditions,
			RevenueTrend:     revenueTrend,
			EmployeeGrowth:   employeeGrowth,
			RiskVolatility:   riskVolatility,
		}
	}

	sdg.logger.Debug("Generated synthetic historical sequence",
		zap.String("business_name", business.BusinessName),
		zap.String("industry", business.Industry),
		zap.Int("months", months),
		zap.Float64("avg_risk_score", sdg.calculateAverageRisk(sequence)))

	return sequence
}

// generateCorrelatedFeature generates a feature that is correlated with the risk score
func (sdg *SyntheticDataGenerator) generateCorrelatedFeature(riskScore, correlation float64, rng *rand.Rand) float64 {
	// Generate a feature that is inversely correlated with risk (lower risk = higher feature value)
	baseValue := 1.0 - riskScore*correlation
	noise := (rng.Float64() - 0.5) * 0.2
	return math.Max(0.0, math.Min(1.0, baseValue+noise))
}

// generateMarketConditions generates market conditions based on time and economic cycles
func (sdg *SyntheticDataGenerator) generateMarketConditions(timestamp time.Time, rng *rand.Rand) float64 {
	// Simulate economic cycles (boom/bust cycles)
	yearsSinceEpoch := timestamp.Year() - 2020
	cyclePhase := math.Sin(float64(yearsSinceEpoch) * 2 * math.Pi / 7) // 7-year cycle

	// Add some randomness
	noise := (rng.Float64() - 0.5) * 0.3

	// Convert to 0-1 range
	marketConditions := (cyclePhase+1)/2 + noise
	return math.Max(0.0, math.Min(1.0, marketConditions))
}

// generateVolatility generates risk volatility based on current risk and industry volatility
func (sdg *SyntheticDataGenerator) generateVolatility(riskScore, industryVolatility float64, rng *rand.Rand) float64 {
	// Higher risk scores tend to have higher volatility
	baseVolatility := industryVolatility * (0.5 + riskScore*0.5)
	noise := (rng.Float64() - 0.5) * 0.1
	return math.Max(0.0, math.Min(1.0, baseVolatility+noise))
}

// calculateAverageRisk calculates the average risk score in a sequence
func (sdg *SyntheticDataGenerator) calculateAverageRisk(sequence []RiskDataPoint) float64 {
	if len(sequence) == 0 {
		return 0.0
	}

	sum := 0.0
	for _, point := range sequence {
		sum += point.RiskScore
	}
	return sum / float64(len(sequence))
}

// initializeIndustryPatterns initializes risk patterns for different industries
func initializeIndustryPatterns() map[string]RiskPattern {
	patterns := make(map[string]RiskPattern)

	// Technology industry
	patterns["technology"] = RiskPattern{
		BaseRisk:            0.25,
		Volatility:          0.15,
		Seasonality:         []float64{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		Trend:               0.02,
		EconomicSensitivity: 0.3,
	}

	// Healthcare industry
	patterns["healthcare"] = RiskPattern{
		BaseRisk:            0.20,
		Volatility:          0.10,
		Seasonality:         []float64{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		Trend:               -0.01,
		EconomicSensitivity: 0.1,
	}

	// Financial services
	patterns["financial"] = RiskPattern{
		BaseRisk:            0.35,
		Volatility:          0.25,
		Seasonality:         []float64{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		Trend:               0.01,
		EconomicSensitivity: 0.8,
	}

	// Manufacturing
	patterns["manufacturing"] = RiskPattern{
		BaseRisk:            0.30,
		Volatility:          0.20,
		Seasonality:         []float64{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		Trend:               0.0,
		EconomicSensitivity: 0.6,
	}

	// Retail
	patterns["retail"] = RiskPattern{
		BaseRisk:            0.40,
		Volatility:          0.30,
		Seasonality:         []float64{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		Trend:               0.0,
		EconomicSensitivity: 0.7,
	}

	// Default pattern
	patterns["default"] = RiskPattern{
		BaseRisk:            0.30,
		Volatility:          0.20,
		Seasonality:         []float64{0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
		Trend:               0.0,
		EconomicSensitivity: 0.5,
	}

	return patterns
}

// GenerateTrainingDataset generates a large dataset for model training
func (sdg *SyntheticDataGenerator) GenerateTrainingDataset(businessCount int, monthsPerBusiness int) map[string][]RiskDataPoint {
	dataset := make(map[string][]RiskDataPoint)

	// Generate data for multiple businesses
	for i := 0; i < businessCount; i++ {
		businessName := fmt.Sprintf("Training_Business_%d", i)
		industry := sdg.getRandomIndustry()

		business := &models.RiskAssessmentRequest{
			BusinessName:    businessName,
			BusinessAddress: fmt.Sprintf("%d Training St", i),
			Industry:        industry,
			Country:         "US",
		}

		sequence := sdg.GenerateHistoricalSequence(business, monthsPerBusiness)
		dataset[businessName] = sequence
	}

	sdg.logger.Info("Generated training dataset",
		zap.Int("business_count", businessCount),
		zap.Int("months_per_business", monthsPerBusiness),
		zap.Int("total_data_points", businessCount*monthsPerBusiness))

	return dataset
}

// getRandomIndustry returns a random industry for training data generation
func (sdg *SyntheticDataGenerator) getRandomIndustry() string {
	industries := []string{"technology", "healthcare", "financial", "manufacturing", "retail"}
	rng := rand.New(rand.NewSource(sdg.randomSeed + time.Now().UnixNano()))
	return industries[rng.Intn(len(industries))]
}

// AddComplianceEvent adds a compliance event to the sequence
func (sdg *SyntheticDataGenerator) AddComplianceEvent(sequence []RiskDataPoint, eventMonth int, severity float64) []RiskDataPoint {
	if eventMonth < 0 || eventMonth >= len(sequence) {
		return sequence
	}

	// Adjust risk score and compliance score for the event month
	sequence[eventMonth].RiskScore = math.Min(1.0, sequence[eventMonth].RiskScore+severity*0.3)
	sequence[eventMonth].ComplianceScore = math.Max(0.0, sequence[eventMonth].ComplianceScore-severity*0.4)

	// Add some spillover effect to adjacent months
	if eventMonth > 0 {
		sequence[eventMonth-1].RiskScore = math.Min(1.0, sequence[eventMonth-1].RiskScore+severity*0.1)
	}
	if eventMonth < len(sequence)-1 {
		sequence[eventMonth+1].RiskScore = math.Min(1.0, sequence[eventMonth+1].RiskScore+severity*0.1)
	}

	return sequence
}

// AddEconomicShock adds an economic shock to the sequence
func (sdg *SyntheticDataGenerator) AddEconomicShock(sequence []RiskDataPoint, shockMonth int, severity float64, duration int) []RiskDataPoint {
	if shockMonth < 0 || shockMonth >= len(sequence) {
		return sequence
	}

	// Apply shock to multiple months
	for i := 0; i < duration && shockMonth+i < len(sequence); i++ {
		monthIndex := shockMonth + i
		decayFactor := 1.0 - float64(i)/float64(duration) // Decay over time

		// Adjust market conditions and risk score
		sequence[monthIndex].MarketConditions = math.Max(0.0, sequence[monthIndex].MarketConditions-severity*decayFactor*0.5)
		sequence[monthIndex].RiskScore = math.Min(1.0, sequence[monthIndex].RiskScore+severity*decayFactor*0.2)
	}

	return sequence
}

// GetIndustryPattern returns the risk pattern for a specific industry
func (sdg *SyntheticDataGenerator) GetIndustryPattern(industry string) (RiskPattern, bool) {
	pattern, exists := sdg.industryPatterns[industry]
	return pattern, exists
}

// SetIndustryPattern sets a custom risk pattern for an industry
func (sdg *SyntheticDataGenerator) SetIndustryPattern(industry string, pattern RiskPattern) {
	sdg.industryPatterns[industry] = pattern
	sdg.logger.Info("Updated industry pattern",
		zap.String("industry", industry),
		zap.Float64("base_risk", pattern.BaseRisk),
		zap.Float64("volatility", pattern.Volatility))
}
